spec/app/ics-101-interchain-swap/README.md

```diff
@@ -0,0 +1,1613 @@
+
+This is an overview of how Interchain Swap works:
+
+![Interchain Swap Diagram](interchain-swap.svg)
+
+### Algorithms
+
+#### Invariant
+
+A constant invariant is maintained after trades, taking into consideration token weights and balance. The value function $V$ is defined as:
+
+$$V = {&Pi;_tB_t^{W_t}}$$
+
+Where
+
+- $t$ ranges over the tokens in the pool
+- $B_t$ is the balance of the token in the pool
+- $W_t$ is the normalized weight of the tokens, such that the sum of all normalized weights is 1.
+
+#### Spot Price
+
+Spot prices of tokens are defined entirely by the weights and balances of the token pair. The spot price between any two tokens, $SpotPrice_i^{o}$, or in short $SP_i^o$, is the ratio of the token balances normalized by their weights:
+
+$$SP_i^o = (B_i/W_i)/(B_o/W_o)$$
+
+- $B_i$ is the balance of token $i$, the token being sold by the trader which is going into the pool
+- $B_o$ is the balance of token $o$, the token being bought by the trader which is going out of the pool
+- $W_i$ is the weight of token $i$
+- $W_o$ is the weight of token $o$
+
+#### Fees
+
+Traders pay swap fees when they trade with a pool. These fees can be customized with a minimum value of 0.01% and a maximum value of 10%. Basis points for `feeRate` is used, where the integer 1 is 0.01% and integer 100 is 1.00%.
+
+The fees go to liquidity providers in exchange for depositing their tokens in the pool to facilitate trades. Trade fees are collected at the time of a swap, and goes directly into the pool, increasing the pool balance. For a trade with a given $inputToken$ and $outputToken$, the amount collected by the pool as a fee is
+
+$$Amount_{fee} = Amount_{inputToken} * swapFee$$
+
+As the pool collects fees, liquidity providers automatically collect fees through their proportional ownership of the pool balance.
+
+### Pool Initialization
+
+An interchain liquidity pool has two possible states:
+
+`POOL_STATUS_INITIAL`: The interchain liquidity pool is created but not yet ready to execute swaps. Pool parameters have been registered but assets for the swap pair have not been deposited.
+
+`POOL_STATUS_READY`: The interchain liquidity pool is ready to execute swaps. Required assets have been deposited.
+
+The pool can be fully funded through the initial deposit or through subsequent deposits. Deposits can be single-asset deposits or multi-asset deposits.
+
+### Data Structures
+
+#### Pool Structure
+
+```ts
+interface Coin {
+  amount: int64;
+  denom: string;
+}
+```
+
+```ts
+enum PoolAssetSide {
+  Source = 1;
+  Destination = 2;
+}
+```
+
+```ts
+// PoolStatus defines if the pool is ready for trading
+enum PoolStatus {
+  INITIALIZED = 0;
+  ACTIVE = 1;
+}
+```
+
+```ts
+interface PoolAsset {
+  side: PoolAssetSide;
+  balance: Coin;
+  // percentage
+  weight: int32;
+  decimal: int32;
+}
+```
+
+```ts
+interface InterchainLiquidityPool {
+  id: string;
+  sourceCreator: string;
+  destinationCreator: string;
+  assets: []PoolAsset;
```

Comment:
This array should always have 2 elements, right? Would it be better for readability and ease of comprehension to actually make two different fields (e.g `sourcePoolAsset` and `destinationPoolAsset`)? Just to avoid working with all these indexes...

Reply:
balancer is for multi-asset deposit. it doesn't mean only 2 tokens. 
of course, currently we only support 2 tokens. 
but in future, we will support several tokens. 
I agree define sourcePoolAsset, or destinationPooAsset, but it will become array.